### DISCUSSION TITLE
Prevent disconnect when there is network issue

### DISCUSSION THREAD

**Discussion by MingChou26082001** [3urobeat/steam-idler#25] [Ideas]
Hi, like the title I have just had an idea where it can prevent disconnect when there is network issue. This helps the program remain reconnecting until the network issue is resolved. I edited a little in bot.js

Bot.prototype.handleRelog = function() {
    if (controller.relogQueue.includes(this.loginindex)) return; // Don't handle this request if account is already waiting for relog

    // Call logPlaytime to print session results and reset startedPlayingTimestamp
    this.logPlaytimeToFile();

    // Add account to queue
    controller.relogQueue.push(this.loginindex);

    // Add a loop for relog
    const retryRelog = async () => {
        let relogSuccess = false;

        while (!relogSuccess) { // Loop until relog is successful
            try {
                // Check if it's our turn to relog every 1 sec after waiting relogDelay ms
                await new Promise(resolve => setTimeout(resolve, config.relogDelay));

                const relogInterval = setInterval(() => {
                    if (controller.relogQueue.indexOf(this.loginindex) != 0) return; // Not our turn? stop and retry in the next iteration

                    clearInterval(relogInterval); // Prevent any retries
                    this.client.logOff();

                    logger("info", `[${this.logOnOptions.accountName}] It is now my turn. Relogging in ${config.loginDelay / 1000} seconds...`);

                    // Attach relog delay timeout
                    setTimeout(async () => {
                        logger("info", `[${this.logOnOptions.accountName}] Logging in...`);

                        // Generate steamGuardCode with shared secret if one was provided
                        if (this.logOnOptions.sharedSecret) {
                            this.logOnOptions.steamGuardCode = SteamTotp.generateAuthCode(this.logOnOptions.sharedSecret);
                        }

                        const refreshToken = await this.session.getToken();
                        if (!refreshToken) return; // Stop execution if getToken aborted login attempt

                        // Call logon
                        this.client.logOn({ "refreshToken": refreshToken });

                    }, config.loginDelay);
                }, 1000);

                // Assure listener only happen once and will be removed when it is done
                const onLoggedOn = () => {
                    logger("info", `[${this.logOnOptions.accountName}] Logged on successfully!`);
                    relogSuccess = true; 
                    // Delete listeners after a success relog
                    this.client.removeListener('error', onError);
                    this.client.removeListener('disconnected', onDisconnected);
                };

                const onError = (err) => {
                    logger("error", `[${this.logOnOptions.accountName}] Error during login: ${err}`);
                    relogSuccess = false;
                    // Delete listeners when error occur
                    this.client.removeListener('loggedOn', onLoggedOn);
                    this.client.removeListener('disconnected', onDisconnected);
                };

                const onDisconnected = (eresult, msg) => {
                    logger("warn", `[${this.logOnOptions.accountName}] Disconnected from Steam: ${msg}`);
                    relogSuccess = false;
                    // Delete listeners if it is disconnected
                    this.client.removeListener('loggedOn', onLoggedOn);
                    this.client.removeListener('error', onError);
                };

                // Event will be assigned once only
                this.client.once('loggedOn', onLoggedOn);
                this.client.once('error', onError);
                this.client.once('disconnected', onDisconnected);

            } catch (error) {
                // Handle error if there is any
                if (error.message.includes("NoConnection") || error.eresult == EResult.NoConnection) {
                    logger("warn", `[${this.logOnOptions.accountName}] Failed to connect. Retrying relog in ${config.retryDelay / 1000} seconds...`);
                    await new Promise(resolve => setTimeout(resolve, config.retryDelay)); 
                } else {
                    logger("error", `[${this.logOnOptions.accountName}] Error during relog: ${error}. Retrying...`);
                    await new Promise(resolve => setTimeout(resolve, config.retryDelay)); 
                }
            }
        }
    };

    retryRelog(); // relog loop
};

**Comment by 3urobeat** [comment]
Hey, I was fairly confident that reconnecting works. If not, then I'd have to check this out again. Are you maybe experiencing an issue on the inital login and not on a relog?

**Reply by MingChou26082001** [reply]
Yes it does work. In some cases that where the network connection is down for less than 15 seconds, it can reconnect without any problems. 
However, if those case are longer than 15 seconds, it will to reconnect for more than 2 times, and if second time retry to connect occurs and still without network, the program will quit instead of retrying the third time and so on.

**Reply by 3urobeat** [reply]
I was just able to replicate that, the bot just exits after reconnecting to the internet.  
I'll let you know when v1.10 is finished which will include a fix.

**Comment by 3urobeat** [comment]
This issue is now fixed and Version 1.10 is out, which can be downloaded from the repository!
